hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs

```diff
@@ -620,3 +635,71 @@ replacePolicyIdWith originalPolicyId otherPolicyId output =
         (AssetId policyId t, q) | policyId == originalPolicyId -> (AssetId otherPolicyId t, q)
         v -> v
    in output{txOutValue = newValue}
+
+replaceSnapshotNumber :: Head.SnapshotNumber -> Head.State -> Head.State
+replaceSnapshotNumber snapshotNumber = \case
+  Head.Closed{parties, utxoHash, contestationDeadline, headId} ->
+    Head.Closed
+      { Head.parties = parties
+      , Head.snapshotNumber = snapshotNumber
+      , Head.utxoHash = utxoHash
+      , Head.contestationDeadline = contestationDeadline
+      , Head.headId = headId
+      }
+  otherState -> otherState
+
+replaceParties :: [Data.Party] -> Head.State -> Head.State
+replaceParties parties = \case
```

Comment:
all head states have parties as an attribute, so i think would be best to rename the builder functions to specify the state that is being updated.

Reply:
Well actually the [`Final`](https://github.com/input-output-hk/hydra/blob/gap/move_attributes_from_redeemer_to_datum_for_close_and_contest_tx/hydra-plutus/src/Hydra/Contract/HeadState.hs#L42) state does not have a `parties` field. That is why we have to pattern match against the type of state that we update.

But other than that, this function `replaceParties` would replace the `parties` field of any given state with does have a `parties` field.

Does that help? Otherwise, maybe you can suggest a code change just to help understand the improvement you have in mind.

Reply:
Oh you are right, we could just extend this function as needed!

Lets go!